### PR TITLE
Added bucket controller

### PR DESCRIPTION
**Pull request by rrati**

### CODE REVIEW

**Comment by rrati** [comment]
wait to merge until this is resolved:
https://github.com/kubernetes-sigs/container-object-storage-interface-spec/issues/12

**Comment by rrati** [comment]
There's a fake client in the spec repo now.

**Review by brahmaroutu — COMMENTED** [review]
Thanks, test cases are comprehensive.
I think we should add some for Anonymous Access mode support.
otherwise LGTM

File: pkg/controller/bucket/bucket_controller.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+		req.AnonymousBucketAccessMode = osspec.ProvisionerCreateBucketRequest_BUCKET_PRIVATE
+	} else if obj.Spec.AnonymousAccessMode.PublicReadOnly {
+		req.AnonymousBucketAccessMode = osspec.ProvisionerCreateBucketRequest_BUCKET_READ_ONLY
+	} else if obj.Spec.AnonymousAccessMode.PublicReadWrite {
+		req.AnonymousBucketAccessMode = osspec.ProvisionerCreateBucketRequest_BUCKET_WRITE_ONLY
+	} else if obj.Spec.AnonymousAccessMode.PublicWriteOnly {
+		req.AnonymousBucketAccessMode = osspec.ProvisionerCreateBucketRequest_BUCKET_READ_WRITE
+	}
+
+	switch obj.Spec.Protocol.Name {
```
[the remark this answers is not in the document]
Comment (rrati) [reply pkg/controller/bucket/bucket_controller.go:118]: That's a good point.  Protocol.Version should probably be added to the BucketContext.

[the remark this answers is not in the document]
Comment (rrati) [reply pkg/controller/bucket/bucket_controller.go:118]: Should probably add SignatureVersion as well for S3

File: pkg/controller/bucket/bucket_controller.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+	case v1alpha1.ProtocolNameS3:
+		req.BucketContext["Region"] = obj.Spec.Protocol.S3.Region
```
[the remark this answers is not in the document]
Comment (rrati) [reply pkg/controller/bucket/bucket_controller.go:120]: It was my understanding that the sidecar would only populate the protocol fields with info it gets from the sidecar.  I can  populate the entire set of protocol fields in the sidecar.  We should discuss where this makes the most sense.  I don't really feel strongly either way.

[the remark this answers is not in the document]
Comment (rrati) [reply pkg/controller/bucket/bucket_controller.go:120]: After discussion, it was decided that bc.parameters will hold all the options needed for the provisioner.  bc.parameters is copied to b.parameters, and this code will send all those parameters to the provisioner.  The provisioner will need to return the data for the protocol fields, and the sidecar will populate all the provisioner specific protocol fields.

Sound right @brahmaroutu ?

File: pkg/controller/bucket/bucket_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rrati) [reply pkg/controller/bucket/bucket_controller.go:118]: Added Protocol.Version, but there's a Protocol.S3.Version that seems conflicting.  For now I have just removed Protocol.S3.Version.

**Comment by rrati** [comment]
Part of #8

**Comment by rrati** [comment]
part of #6

File: pkg/controller/bucket/bucket_controller.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+// NewBucketController returns a controller that manages Bucket objects
+func NewBucketController(provisionerName string, client osspec.ProvisionerClient) (*controller.ObjectStorageController, error) {
+	rateLimit := workqueue.NewMaxOfRateLimiter(
+		workqueue.NewItemExponentialFailureRateLimiter(5*time.Second, 60*time.Minute),
+		&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(10), 100)},
+	)
+
+	identity := fmt.Sprintf("object-storage-sidecar-%s", provisionerName)
```
Comment (brahmaroutu) [inline pkg/controller/bucket/bucket_controller.go:62]: I think we would want to standardize the names objectstorage-sidecar-XXXX

File: pkg/controller/bucket/bucket_controller.go [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+func (bl *bucketListener) InitializeBucketClient(bc bucketclientset.Interface) {
+	bl.bucketClient = bc
+}
+
+// Add will call the provisioner and add a bucket
+func (bl *bucketListener) Add(ctx context.Context, obj *v1alpha1.Bucket) error {
+	klog.V(1).Infof("bucketListener: add called for bucket %s", obj.Name)
+
+	// Verify this bucket is for this provisioner
+	if !strings.EqualFold(obj.Spec.Provisioner, bl.provisionerName) {
```
Comment (brahmaroutu) [inline pkg/controller/bucket/bucket_controller.go:99]: who provides the provisionerName for the Sidecar?

Comment (rrati) [reply pkg/controller/bucket/bucket_controller.go:99]: The sidecar queries the driver for that information.  It's in code to be merged after this is in.

File: pkg/controller/bucket/bucket_controller.go (hunk 3 above)
Comment (rrati) [reply pkg/controller/bucket/bucket_controller.go:62]: I'll do that when I push up the main binary bits.

**Pull request merged**